File: Java/src/_0980_Unique_Paths_III/Solution.java
package _0980_Unique_Paths_III;

/**
 * 执行用时：2 ms, 在所有 Java 提交中击败了21.70%的用户
 * 内存消耗：35.7 MB, 在所有 Java 提交中击败了61.60%的用户
 */
public class Solution {

    private int[][] grid;
    private int n, m;
    private boolean[][] st;
    private int start, end;

    private final static int[][] dirs = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};  // 某个点的四个方向，上右下左

    public int uniquePathsIII(int[][] grid) {

        this.grid = grid;
        n = grid.length;
        m = grid[0].length;
        st = new boolean[n][m];

        int cnt = n * m;  // 初始表示多少个可以走的格子数目
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
                if (grid[i][j] == 1) start = i * m + j;
                else if (grid[i][j] == 2) end = i * m + j;
                else if (grid[i][j] == -1) cnt--;

        return dfs(start, cnt);
    }

    private int dfs(int u, int cnt) {

        int x = u / m, y = u % m;
        if (--cnt == 0 && u == end) {
            st[x][y] = false;
            return 1;
        }

        int res = 0;
        st[x][y] = true;
        for (int d = 0; d < 4; d++) {
            int a = x + dirs[d][0], b = y + dirs[d][1];
            if (a >= 0 && a < n && b >= 0 && b < m && grid[a][b] != -1 && !st[a][b])
                res += dfs(a * m + b, cnt);
        }
        st[x][y] = false;

        return res;
    }

    public static void main(String[] args) {

        int[][] grid = {
                {1, 0, 0, 0},
                {0, 0, 0, 0},
                {0, 0, 0, 2}
        };
        System.out.println((new Solution()).uniquePathsIII(grid));  // 4
    }
}

File: Python/_0980_Unique_Paths_III/Solution.py
# coding=utf-8
# Date: 2021/7/1 14:40
from typing import List


# 执行用时：100 ms, 在所有 Python3 提交中击败了20.00%的用户
# 内存消耗：15.2 MB, 在所有 Python3 提交中击败了10.94%的用户
class Solution:
    def __init__(self):
        self.n, self.m = 0, 0
        self.st = []
        self.S, self.T = [0, 0], [0, 0]
        self.ans = 0
        self.dx, self.dy = [-1, 0, 1, 0], [0, 1, 0, -1]

    def uniquePathsIII(self, grid: List[List[int]]) -> int:
        self.n, self.m = len(grid), len(grid[0])
        self.st = [[False for _ in range(self.m)] for _ in range(self.n)]
        self.ans = 0

        cnt = self.n * self.m
        for i in range(self.n):
            for j in range(self.m):
                if grid[i][j] == 1:
                    self.S = [i, j]
                elif grid[i][j] == 2:
                    self.T = [i, j]
                elif grid[i][j] == -1:
                    cnt -= 1
        self.dfs(grid, self.S[0], self.S[1], cnt - 1)
        return self.ans

    def dfs(self, grid, x, y, cnt):

        if cnt == 0 and x == self.T[0] and y == self.T[1]:
            self.ans += 1
            return

        self.st[x][y] = True
        for i in range(4):
            a, b = x + self.dx[i], y + self.dy[i]
            if 0 <= a < self.n and 0 <= b < self.m and not self.st[a][b] and grid[a][b] != -1:
                self.dfs(grid, a, b, cnt - 1)
        self.st[x][y] = False


if __name__ == "__main__":
    print(Solution().uniquePathsIII([[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 2, -1]]))
